Update report format to ga4gh-testbed-lib

alextsaihi opened the pull request:
testbed-lib json good

still unable to serve report as html
jinja2.exceptions.UndefinedError: 'str object' has no attribute 'server_name'

alextsaihi committed:
Update cli.py

alextsaihi committed:
just added platform name stuff

just added platform name stuff

alextsaihi committed:
-r flag

-r to output rnaget report format

alextsaihi committed:
Update cli.py

jb-adams commented:
Hi @saupchurch and @emi80, Alex Tsai (@alextsaihi ) is a new developer on my team. His first task has been to update the RNAget compliance suite, so that it outputs a report according to a common [ga4gh-testbed-lib](https://github.com/ga4gh/ga4gh-testbed-lib) format. The same work is also being done in the refget compliance suite, so our aim is to have both test suites use the same report format.

Can we get your feedback on the following points:
- How does this plan sound to you?
- @alextsaihi added a command line flag so that the user can choose to either output things according to the old format or new format. Should we aim to preserve the old report format, or scrap it altogether for the simplicity and interoperability of only supporting the new format?
- The RNAget mini server will obviously not be able to render reports according to the new format. Our plan is to develop an API/website where reports from many different GA4GH testbeds can be `POSTed` to for aggregation and visualization. With this being said, should we (a) scrap the RNAget mini server in favour of the new API in development, or (b) update the mini server to enable rendering of reports by the new format?

@Chen2X @yash-puligundla looping you in as this will have ramifications for Refget and the testbed API

jb-adams commented:
Thanks Sean, I appreciate the feedback!

I'm leaning toward removing the old report format and the flag, and just supporting the new format going forward.

I'm also thinking that we should update the mini server/UI in RNAget to support the new format. If we take this out, then someone who wants to "kick the tires" and test their server will have to register on the central API just to view the report. This is a bit overkill for quick testing.

emi80 commented:
Hi @jb-adams and @saupchurch,
I completely agree with what have been said so far. 

About the RNAget mini server, I am in favor of updating it to support the new format if that sounds reasonable. I am also available for giving some help with developing and/or testing in case it's needed.

alextsaihi committed:
some more additions

--pretty flag, case messages, case status updates

alextsaihi commented:
Hi @saupchurch and @emi80 , could you guys check if Caltech server is up? I was testing against the Caltech RNAget server and it seems to be down at the moment.

saupchurch commented:
The Caltech RNAget server is currently down - there are some issues unrelated to the RNAget instance that we are working to resolve

alextsaihi committed:
update 03/08/2022

alextsaihi committed:
JSON end-to-end testing for testbed format

currently have port numbers for mock server changed to 8000 from 5000 because of local machine issues

alextsaihi committed:
reverted port number changes

alextsaihi committed:
updated test summary/message

emi80 commented:
Hi @alextsaihi,

I managed to make a test run of your branch against the CRG server and it seems everything worked fine. Just a few comments:

1. some documentation update is needed, both on installing the suite and on the configuration
    * the `python setup.py install` command did not work for me (with `Python 3.10`) and I had to install the suite with `pip install .`
    * we have been using configuration files potentially containing a list of servers so far, while now it seems only one server object is supported
2. if I understand correctly, only the testing framework has been changed (using `ga4gh-testbed-lib`) while there is no change on the HTML report output. Please correct me if I am wrong.

Thanks a lot for your work.

Best

saupchurch commented:
I'm having issues installing the ga4gh module.  From the error it looks like it hasn't been updated for Python 3 (my environment is using Python 3.9.1):
scripts/process_schemas.py line 94
          except OSError, e:
                        ^
      SyntaxError: invalid syntax
      [end of output]

emi80 commented:
Hi Sean,
I guess you are trying to install the `ga4gh` python module, but that's an old module related to the former reference implementation. Just install `ga4gh-testbed-lib` and you should be good to go.

It's true that the error message you get if the testbed lib is missing is a bit misleading as it refers to the module `ga4gh`... I think that should be fixed.

saupchurch commented:
Thanks @emi80 that did the trick.  The test suite runs now.  I second the concerned addressed above:
1) documentation needs to be updated
      - python setup.py install should be fixed
      - requirements_test.txt needs to be reviewed and updated
      - it seems like the config.yml is only supporting one server because we are moving to a "run yourself then submit results" model.  If so this should be reflected in updated documentation

Looking good so far!

emi80 commented:
@saupchurch, you are welcome! I also agree with all your comments.

alextsaihi committed:
separated config implementation .yaml into separate servers

the goal was to only run against one single server at a time

alextsaihi commented:
Hi @emi80 , I have tried to reproduce the installation error you were receiving on a new venv running python 3.10.2, but with no success. Could you let me know what error you were receiving from `python setup.py install`?

alextsaihi committed:
README.md and dependencies fix

emi80 commented:
Hi @alextsaihi,

thanks for the updates, 80c631e looks good. Just a couple of comments: 

* I would probably use `numpy<1.24,>=1.18` as the requirement filter so that you can get the latest supported `numpy` version

* I just noticed that if you put the same requirement in `setup.py` you can just do `python setup.py install`, without the need to previously install the requirements with `pip`:

  ```python
  ...
  install_requires = ['requests', 'click', 'jsonschema', 'PyYAML', "Jinja2", "numpy<1.24,>=1.18", "loompy"] 
  ...
  ```
  Since keeping versions in sync between `setup.py` and requirements files can be difficult to maintain, we might also separate requirements into different files and do something like this in `setup.py`:
  
  ```python
  with open("requirements_setup.txt", "r") as fh:                                                                                                                                                                                                  
      install_requires = fh.read().splitlines()
  ```

What do you think @alextsaihi @saupchurch? I'd suggest to make the two changes above and then we can merge the PR.

Best,
Emilio

saupchurch commented:
Good points @emi80 .  I'm not sure exactly how I feel about allowing the most recent versions of required software.  I've been burned in the past when elements get deprecated and a new install is now broken.  It's something that can be managed, but would put us in the position of having to chase and test with version releases of dependencies.

As far as the second comment I think this is a great concept: having requirements in two locations is not an ideal design and we should decide on a single place to have them.

alextsaihi commented:
Since numpy versions 1.23.x do exist I would agree to have a <1.24.0 upper bound for the numpy version check. (I'll have to test if numpy 1.23.5 is compatible with 'numba')

I have tried replacing the install_requires variable with a list straight from requirements_doc.txt, but I'm running into some errors that probably needs debugging (I've made sure to split out the "==ver.number" out of each line). I could spend some more time tomorrow to figure this out.

error: Could not find suitable distribution for Requirement.parse('urllib3==1.25.3')

code:

with open("requirements_doc.txt", "r") as fh:                                                                                                                                                                                                  
    install_requires = fh.read().splitlines()

for i in install_requires:
    i = re.split("[=><]", i)[0]

alextsaihi commented:
I have just done some reading on the current convention to keep setup.py and requirements.txt separate:

https://stackoverflow.com/questions/14399534/reference-requirements-txt-for-the-install-requires-kwarg-in-setuptools-setup-py

"So the takeaway:

setup.py should declare the loosest possible dependency versions that are still workable. Its job is to say what a particular package can work with.

requirements.txt is a deployment manifest that defines an entire installation job, and shouldn't be thought of as tied to any one package. Its job is to declare an exhaustive list of all the necessary packages to make a deployment work.

Because these two things have such different content and reasons for existing, it's not feasible to simply copy one into the other."

Currently some of the other compliance suites also has the two dependencies checks separate and it could be a good idea to leave it as is for rnaget as well. What do you guys think?

emi80 commented:
Thanks @alextsaihi for taking the time to have a look at this!

Following the stackoverflow thread, I think it makes sense to keep the two lists of dependencies separated. 

By the way, I just found out that we can have loose dependencies in `setup.py` and, at the same time, make `python setup.py install` work without issues by setting the following for `install_requires`:

```
...
install_requires = ['requests', 'click', 'jsonschema', 'PyYAML', "Jinja2", "loompy", "numba"]
...
```

At least it works for me...

alextsaihi committed:
dependencies adjustments (numba)

alextsaihi commented:
Thats a good suggestion, I've just added numba to the install_requires and made the edits to requirements_doc.txt to include the 1.23.x versions of numpy. I've tested with numpy 1.23.5 and it seems to work fine.

emi80 commented on README.md:
````diff
@@ -21,6 +21,13 @@ RNAget maintains an up-to-date report of the compliance status of its reference
+### Installation
+
+```
+pip install -r requirements_doc.txt
````
@alextsaihi I think we can probably revert this back now that `install_requires` is updated

emi80 replied:
HI @alextsaihi. Yes, I think that now `python setup.py install` should just work.

emi80 replied:
Addressed in 1f41339

emi80 commented on setup.py:
```diff
@@ -17,7 +18,8 @@
     long_description = fh.read()
-install_requires = ['requests', 'click', 'jsonschema', 'PyYAML', "Jinja2", "loompy"]
+
+install_requires = ['requests', 'click', 'jsonschema', 'PyYAML', "Jinja2", "loompy", "numba"]
```
We also need `ga4gh-testbed-lib` here otherwise we get `ModuleNotFoundError: No module named 'ga4gh'` after `python setup.py install`

emi80 commented:
Hi @alextsaihi, I made two (hopefully last) comments to the PR. 
I actually installed and run the compliance test suite against the CRG demo server and it worked, so I think we can merge the PR once these two comments are resolved.

Thanks!

alextsaihi committed:
ga4gh-testbed-lib in install_requires

alextsaihi commented:
sounds good! those changes have just been pushed
